### PR TITLE
Adding support for logging additional artifacts when using mfllow pytorch log and save model

### PR DESCRIPTION
**Pull request by shrinath-suresh**
## What changes are proposed in this pull request?

In the current version of mlflow pytorch log/save model, there is no mechanism available to push artifact files along with the pytorch model. The supporting files such as requirements.txt or vocabulary files would be really helpful in case of deployments or new environment setups. Hence, adding a new parameter `artifacts` to mlflow pytorch log and save model which would help user to log supporting files into mlflow.

## How is this patch tested?

Tested the code by logging artifact files along with log and save model. And also ensured that the existing scenario is working as expected(saving and logging model without artifact parameter). Added unit test cases as well. 

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [x] Yes. Give a description of this change to be included in the release notes for MLflow users.

Users can log additional artifact files such as requirements, vocabulary files along with the pytorch model when using `mlflow.pytorch.log_model` or `mfllow.pytorch.save_model` by adding them in artifacts parameter. 

### What component(s), interfaces, languages, and integrations does this PR affect?
Components 
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [x] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: Local serving, model deployment tools, spark UDFs
- [ ] `area/server-infra`: MLflow server, JavaScript dev server
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface 
- [ ] `area/uiux`: Front-end, user experience, JavaScript, plotting
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language 
- [ ] `language/r`: R APIs and clients
- [x] `language/python`: Python APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

**Comment by harupy** [comment]
Why not use `mlflow.log_artifact` or `log_artifacts`?

**Comment by shrinath-suresh** [comment]
> @shrinath-suresh
> Why not use `mlflow.log_artifact` or `mlflow.log_artifacts`?

The change is very specific to pytorch models which would facilitate bundling of all the pytorch related artifacts along with the model. This would be complex when we use `mlflow.log_artifact`. The earlier method would also facilitate  easy reading of the artifacts during deployment of the pytorch model.

**Comment by shrinath-suresh** [comment]
@smurching Looks like there is an existing pylint issue in the code.

1. https://github.com/mlflow/mlflow/blob/64a51f43a64e91891a23985a9096a4dc3d1d6150/tests/pytorch/test_pytorch_model_export.py#L106 - Eventhough forward function is implemented, pylint is throwing `W0223: Method '_forward_unimplemented' is abstract in class 'Module' but is not overridden (abstract-method)` . Is it okay to disable the warning using - `# pylint: disable=W0223`

2. https://github.com/mlflow/mlflow/blob/64a51f43a64e91891a23985a9096a4dc3d1d6150/mlflow/pytorch/__init__.py#L380 - `Consider explicitly re-raising using the 'from' keyword (raise-missing-from)` . Should we do something like `raise MlflowException("Exception Message") from None`.

File: mlflow/pytorch/__init__.py [hunk 1]
```diff
@@ -318,6 +360,25 @@ def save_model(
     pickle_module_path = os.path.join(model_data_path, _PICKLE_MODULE_INFO_FILE_NAME)
     with open(pickle_module_path, "w") as f:
         f.write(pickle_module.__name__)
+
+    if artifacts:
+        if not isinstance(artifacts, dict):
+            raise TypeError("Argument artifacts should be a list")
```
Comment (nfx) [inline mlflow/pytorch/__init__.py:366]: ... logical typo: `should be a dict`

File: mlflow/pytorch/__init__.py [hunk 2]
```diff
@@ -318,6 +360,25 @@ def save_model(
...
+            saved_artifacts_dir_subpath = "artifacts"
+            for artifact_name, artifact_uri in artifacts.items():
+                tmp_artifact_path = _download_artifact_from_uri(
+                    artifact_uri=artifact_uri, output_path=tmp_artifacts_dir.path()
+                )
+                tmp_artifacts_config[artifact_name] = tmp_artifact_path
+
+            shutil.move(
+                tmp_artifacts_dir.path(),
+                os.path.join(path, saved_artifacts_dir_subpath),
```
Comment (nfx) [inline mlflow/pytorch/__init__.py:379]: from what i see in other reviews, `posixpath` is more frequently used in this repository, than `os.path.join`

File: mlflow/pytorch/__init__.py [hunk 3]
```diff
@@ -366,7 +427,10 @@ def _load_model(path, **kwargs):
         pickle_module_path = os.path.join(path, _PICKLE_MODULE_INFO_FILE_NAME)
         with open(pickle_module_path, "r") as f:
             pickle_module_name = f.read()
-        if "pickle_module" in kwargs and kwargs["pickle_module"].__name__ != pickle_module_name:
+        if (
```
Comment (nfx) [inline mlflow/pytorch/__init__.py:430]: please consider putting formatting changes into separate PR, because they increase the scope of the review.

Comment (shrinath-suresh) [reply mlflow/pytorch/__init__.py:430]: sure i will send a separate PR for formatting changes. Thanks.

**Comment by shrinath-suresh** [comment]
@nfx Accidentally pushed the change from master and created this pull request. I have few other PRs to be pushed. Closing this PR. will raise a newer one from feature branch.

**Comment by shrinath-suresh** [comment]
https://github.com/mlflow/mlflow/pull/3407

**Pull request closed without merging**